### PR TITLE
Improve globe atmosphere rendering

### PR DESCRIPTION
**Pull request by mpulkki-mapbox**
Prior to this change globe's gradient atmosphere effect was achieved by first computing the center point of the globe in screen coordinates and then interpolating the color ramp for each pixel by measuring their distances from the center. This was a very crude approximation which worked okayish with opaque styles as there was plenty of padding between the surface of the globe and the starting point of the gradient effect. Problems of this approach becomes clearly visible with styles using transparent background.

This PR changes the logic a bit by computing pixel distances from the center in view space instead of screen space. I chose a small padding of 1.0 pixels so that there's no background pixels bleeding outside due to non-conservative rendering. I used iPhone XS, Amazon Fire HD tablet and my Dell XPS laptop for testing the change.

In the following images black color is the background seen through the transparent ground. Left is prior and right is post this change.
image

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] include before/after visuals or gifs if this PR includes visual changes
 - [x] manually test the debug page
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'

### CODE REVIEW

File: src/render/draw_globe_atmosphere.js [hunk 1]
```diff
@@ -19,26 +20,37 @@ function drawGlobeAtmosphere(painter: Painter) {
...
     const viewToProj = transform._camera.getCameraToClipPerspective(transform._fov, transform.width / transform.height, transform._nearZ, transform._farZ);
-    const globeToView = mat4.mul([], viewMatrix, transform.globeMatrix);
-    const viewToScreen = mat4.mul([], transform.labelPlaneMatrix, viewToProj);
+    const projToView = mat4.invert([], viewToProj);
 
-    const centerOnViewSpace = vec3.transformMat4([], [0, 0, 0], globeToView);
-    const radiusOnViewSpace = vec3.add([], centerOnViewSpace, [transform.worldSize / Math.PI / 2.0, 0, 0]);
+    const project = (point: Vec3, m: Mat4): [number, number, number] => {
+        vec3.transformMat4(point, point, m);
+        return [point[0], point[1], point[2]];
```
Comment (karimnaaji) [inline src/render/draw_globe_atmosphere.js:32]: nit: We should be able to directly `return vec3.transformMat4(point, point, m);` here, it mutates the input `point` variable anyway, so I don't think we need the extra array:
```suggestion
        return vec3.transformMat4(point, point, m);
```

**Review by karimnaaji — APPROVED** [review]
Works great, I have one nit but otherwise good to merge.

File: src/render/draw_globe_atmosphere.js (hunk 1 above)
Comment (mpulkki-mapbox) [reply src/render/draw_globe_atmosphere.js:32]: I actually spent way more time on this function than I wanted to. The problem here is that `vec3.transformMat4()` expects the input coordinate to have a type of `Vec3` which is defined as `type VecType = Array<number> | Float32Array | Float64Array;`. This type will eventually need to be converted/casted to a tuple in order to be passed to shaders as a uniform. Flow seems to be really strict about this as a simple type cast between `Array<number>` and `[number, number, number]` is not allowed.

I tried quite a many different type permutations but couldn't find any better than this one. I'm quite sure there's a proper way doing this with flow that I'm not aware of (cc @mourner).

Comment (mpulkki-mapbox) [reply src/render/draw_globe_atmosphere.js]: thanks @mourner!

**Pull request merged**